### PR TITLE
Ambianic Edge installation on PWA

### LINKED ISSUE #380: Full Install via UI PWA - Improved app install flow
Based on user feedback, we are revisiting the app install flow.
[New state diagram](https://app.diagrams.net/?lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1&title=Ambianic.ai%20Install%20Flow#Uhttps%3A%2F%2Fdrive.google.com%2Fuc%3Fid%3D14LJfaOkwoWU_HZgIsVgqUuX2g9-9Nk8l%26export%3Ddownload)

[New UI flow mockup](https://app.diagrams.net/?lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1&page-id=igIhIKNUZmc0qu4BMqyh&title=Ambianic.ai%20Install%20Flow#Uhttps%3A%2F%2Fdrive.google.com%2Fuc%3Fid%3D14LJfaOkwoWU_HZgIsVgqUuX2g9-9Nk8l%26export%3Ddownload)

[Diagram source](https://app.diagrams.net/#G14LJfaOkwoWU_HZgIsVgqUuX2g9-9Nk8l)

### PR DESCRIPTION
**Pull request by vickywane**
This closes #380 

This PR contains work done to improve the Ambianic Edge installation using the existing Vue PWA. 

While this pull request is still in progress, the following have been completed;
- Basic installation flow flowing the link sketch [here](https://viewer.diagrams.net/?highlight=0000ff&edit=_blank&layers=1&nav=1&page-id=igIhIKNUZmc0qu4BMqyh&title=Ambianic.ai%20Install%20Flow#Uhttps%3A%2F%2Fdrive.google.com%2Fuc%3Fid%3D14LJfaOkwoWU_HZgIsVgqUuX2g9-9Nk8l%26export%3Ddownload)
- Setting up the messaging clients
- Persisting the installation state on PWA using `localStorage`

### CODE REVIEW

**Comment by vickywane** [comment]
@ivelin I would rebase and squash my commits when I am done with this. 

I also had to reduce the noise ( logs )  in the browser console by commenting them out alongside eslint. I would restore these back after development.

**Comment by ivelin** [comment]
@vickywane converting this PR to draft status. CI checks need to all pass before its ready for review and merge.

**Comment by ivelin** [comment]
@vickywane I see 14 files changed. I suggest keeping PRs small so they are easier and quicker to review and merge. See [this best practice guide](https://www.netlify.com/blog/2020/03/31/how-to-scope-down-prs/).

**Comment by vickywane** [comment]
I would restore most of the files back to what they were previously.

**Comment by ivelin** [comment]
@vickywane could you please add cypress test to cover the new install flow. Thank you!

**Comment by ivelin** [comment]
@vickywane could you please address the LGTM and commit-lint flags.

**Comment by ivelin** [comment]
@vickywane looks like there are a few merge conflicts with the main branch. See comment above.

**Comment by vickywane** [comment]
I would look into covering with cypress. Wasn't expecting to, but I would look into it. What exact things would do you want covered?

**Comment by ivelin** [comment]
Just following the typical code coverage best practices. Focus on covering
the new code introduced in this PR. Thank you!

Codecov will help you see what lines may not be covered yet when you run
the tests with
```
npm run test
```

See report here. We clearly need to improve coverage.
https://codecov.io/gh/ambianic/ambianic-ui

**Comment by ivelin** [comment]
@vickywane thanks for the new commits. It seems like a number of CI checks still fail. Let's keep this PR in draft status until all checks are green. Thank you!

**Comment by vickywane** [comment]
Okay. 

I am still rounding up on the cypress tests, all should be ready tonight.

**Comment by ivelin** [comment]
@vickywane here is a screencast with my install. Specifically the issue with discovering a local device. https://youtu.be/-JUnMDoGlIo

**Comment by vickywane** [comment]
Thank you for the detailed report!

I would work on that flow and test with John's device.

**Comment by vickywane** [comment]
Hello @ivelin , 

The issue has been fixed and cypress tests added. I am now left with getting these checks to a green state. 
Have you retried the issue on your devices?

**Comment by ivelin** [comment]
> Hello @ivelin ,
> 
> The issue has been fixed and cypress tests added. I am now left with getting these checks to a green state.
> Have you retried the issue on your devices?

@vickywane I still observe the same issue. Sees screencast: https://drive.google.com/file/d/10l1qlypUPL-rWbmOhwXa1djkXxFtAQmG/preview

**Comment by ivelin** [comment]
> > Hello @ivelin ,
> > The issue has been fixed and cypress tests added. I am now left with getting these checks to a green state.
> > Have you retried the issue on your devices?
> 
> @vickywane I still observe the same issue. See screencast: https://drive.google.com/file/d/10l1qlypUPL-rWbmOhwXa1djkXxFtAQmG/preview

I see now that your pre-deploy check failed which means that staging may be running your previous version of the code.

**Comment by vickywane** [comment]
@ivelin i seem to have something breaking a new build on Netlify. I'll fix that and inform you for a new test

File: src/views/Home.vue [hunk 1]
```diff
@@ -1,101 +1,99 @@
...
+  </div>
 </template>
 
 <script>
-import UpdateNotification from '@/components/UpdateNotification'
-
 export default {
   name: 'Home',
-  components: {
-    UpdateNotification
```
Comment (ivelin) [inline src/views/Home.vue:98]: @vickywane why did you get rid of the updatenotification? It is what allows us to push new PWA versions to production with the correct UX.

Comment (vickywane) [reply src/views/Home.vue:98]: I thought it wasn't being used. I would restore it back

**Comment by ivelin** [comment]
> @ivelin i seem to have something breaking a new build on Netlify. I'll fix that and inform you for a new test

Right. And that is probably why , I can't test your local edge device discovery fix. The old version of your code is on staging.

File: src/views/Home.vue (hunk 1 above)
Comment (ivelin) [reply src/views/Home.vue:98]: It is used all the time. The user gets a friendly pop-up that their app has a new version. They can click refresh when they are ready to load it. It's been working nicely for quite some time. 

Which reminds me that we need to cover it with tests so developers get a flag if they remove it again or mess it up. I'll open a separate issue for that. Not part of this PR.

**Comment by vickywane** [comment]
> > @ivelin i seem to have something breaking a new build on Netlify. I'll fix that and inform you for a new test
> 
> Right. And that is probably why , I can't test your local edge device discovery fix. The old version of your code is on staging.

What was breaking has been cleared out. I was hoping that would fix the breaking lighthouse and CI here but it didn't :(

You can test again. I believe it should be fixed.

**Comment by ivelin** [comment]
> > > @ivelin i seem to have something breaking a new build on Netlify. I'll fix that and inform you for a new test
> > 
> > 
> > Right. And that is probably why , I can't test your local edge device discovery fix. The old version of your code is on staging.
> 
> What was breaking has been cleared out. I was hoping that would fix the breaking lighthouse and CI here but it didn't :(
> 
> You can test again. I believe it should be fixed.

From what I see, it's still not fixed. And there are other issues to address with install state for first time and return users. See screencast:
https://drive.google.com/file/d/1Z6XJcZrD1t4RF-73AWuwcaDe_CtWXj0P/preview

I suggest a more thorough analysis of the issues at hand. Make sure all CI checks pass and test coverage of new code is thorough before requesting another PR review.

Here is a reference to the project contribution guidelines: https://docs.ambianic.ai/legal/CONTRIBUTING/

**Comment by vickywane** [comment]
@ivelin I've had a thorough look at the problem and I can confirm it's fixed now. 

 Would you prefer some sort of message is displayed for users returning to the installation flow and what words would they be?

File: cypress/integration/ambianic-tests/settings.spec.js [hunk 2]
```diff
@@ -2,8 +2,8 @@
 
 context('Settings', () => {
     before(() => {
-      cy.visit('http://localhost:8080')
-      cy.get('[data-cy=settings]').click()
+      cy.visit('http://localhost:8080/settings')
```
Comment (ivelin) [inline cypress/integration/ambianic-tests/settings.spec.js:5]: @vickywane I see the click() action has been removed. Is this test useful without any assertion?

File: cypress/integration/ambianic-tests/timeline.spec.js [hunk 3]
```diff
@@ -3,7 +3,8 @@
 context('Timeline', () => {
     before(() => {
       cy.visit('http://localhost:8080')
-      cy.get('[data-cy=timeline]').click()
+    //   cy.get('[data-cy=timeline]').click()
```
Comment (ivelin) [inline cypress/integration/ambianic-tests/timeline.spec.js:6]: @vickywane same comment here. What is the point of tests if their assertions are commented out?

**Comment by ivelin** [comment]
@vickywane good progress on test coverage and CI check. Walking through the install, I see a few items that need attention. Screenshots and comments follow.

**Comment by ivelin** [comment]
Mobile users are our primary target. Horizontal stepper still takes only 1/3 of the screen, leaving 2/3 white space, which is not an optimal UX. This leads to other issues in the following install steps. See screenshots below.

Screen Shot 2021-01-20 at 2 45 06 PM

**Comment by ivelin** [comment]
Local edge discovery button is now synced with internal peer connection state. However the UX seems a bit geeky. There is a lot of verbiage and a cryptic peer ID, which doesn't mean much to most users. Can we do something on this page to be in line with user expectations using popular consumer apps? Maybe a green check mark indicating successful connection and an info button that lets advanced users reveal the peer id for debugging purpose as needed.

Screen Shot 2021-01-20 at 2 46 01 PM

**Comment by ivelin** [comment]
This page is not mobile friendly. Some buttons aren't visible. 

Screen Shot 2021-01-20 at 2 46 11 PM

**Comment by ivelin** [comment]
Once the size of the window is expanded sufficiently, the buttons are visible. However this page probably doesn't need to show Reset and Share buttons as it confuses the install flow. These buttons make more sense in the standalone device settings page that users can open later on after install is finished. Additionally these buttons simply don't work on this page.

Screen Shot 2021-01-20 at 2 46 32 PM

Screen Shot 2021-01-20 at 2 46 59 PM

Screen Shot 2021-01-20 at 2 47 04 PM

**Review by ivelin — CHANGES_REQUESTED** [review]
Good progress. Posted a few comments on code changes and UX. Please take a look and address.

**Comment by sniperadmin** [comment]
Great job so far!
I will have a look and see what I can do for enhancing the lighthouse performance :)
I am doubting one or two packages are eating it

File: cypress/integration/ambianic-tests/settings.spec.js (hunk 2 above)
Comment (vickywane) [reply cypress/integration/ambianic-tests/settings.spec.js:5]: The setting button no longer exists on the home page due to the new installation flow. 

The purpose of the click action was to navigate to the settings page which using `visit` does. 

@ivelin this also applies to similar test suites that had to do with the previous Home screen.

**Comment by vickywane** [comment]
The UI changes requested have been made.

@ivelin now that the tests have passed, can I convert this to a PR in line for merging?

I had to resolve a merge conflict due to working from different devices and this is preventing me from rewording my initial commit in a rebase for the `commit-lint-pr` check to pass.

**Comment by vickywane** [comment]
> Once the size of the window is expanded sufficiently, the buttons are visible. However this page probably doesn't need to show Reset and Share buttons as it confuses the install flow. These buttons make more sense in the standalone device settings page that users can open later on after install is finished. Additionally these buttons simply don't work on this page.
> 
> Screen Shot 2021-01-20 at 2 46 32 PM
> 
> Screen Shot 2021-01-20 at 2 46 59 PM
> 
> Screen Shot 2021-01-20 at 2 47 04 PM

This is a valid observation. 

If we would be using only the **OK** button, can we rename it to **Complete Installation** or something similar?

File: cypress/integration/ambianic-tests/settings.spec.js (hunk 2 above)
Comment (ivelin) [reply cypress/integration/ambianic-tests/settings.spec.js:5]: OK, makes sense. Closing.

File: cypress/integration/ambianic-tests/timeline.spec.js (hunk 3 above)
Comment (ivelin) [reply cypress/integration/ambianic-tests/timeline.spec.js:6]: @vickywane 
We should have some sort of assertion test. 
Should this be a visit to the /timeline URL?
And what are we expecting to see once the page loads?

**Comment by ivelin** [comment]
> > Once the size of the window is expanded sufficiently, the buttons are visible. However this page probably doesn't need to show Reset and Share buttons as it confuses the install flow. These buttons make more sense in the standalone device settings page that users can open later on after install is finished. Additionally these buttons simply don't work on this page.
> 
> This is a valid observation.
> 
> If we would be using only the **OK** button, can we rename it to **Complete Installation** or something similar?

Yes, **Complete Installation** seems sensible at this step. Also Skip should not appear in the upper right corner because it makes no sense at this step.

**Comment by ivelin** [comment]
> @ivelin I've had a thorough look at the problem and I can confirm it's fixed now.
> 
> Would you prefer some sort of message is displayed for users returning to the installation flow and what words would they be?

@vickywane Anything that's reasonable and intuitive.

File: cypress/integration/ambianic-tests/settings.spec.js (hunk 2 above)
Comment (ivelin) [reply cypress/integration/ambianic-tests/settings.spec.js:5]: @vickywane actually, I agree with the change to visit /settings, but we still need to make some sort of reasonable assumption about the contents of the settings page.

Comment (vickywane) [reply cypress/integration/ambianic-tests/settings.spec.js:5]: I think the code coverage of the setting page is out of the scope of this PR.

File: cypress/integration/ambianic-tests/timeline.spec.js (hunk 3 above)
Comment (vickywane) [reply cypress/integration/ambianic-tests/timeline.spec.js:6]: That's a before hook event not really an assertion. It's what happens the first time the test case is run. In this case to visit the `timeline` page.

**Review by ivelin — APPROVED** [review]
Great work!

File: src/views/Home.vue [hunk 4]
```diff
@@ -90,12 +99,83 @@
...
+      peerConnectionStatus: (state) => state.pnp.peerConnectionStatus,
+      isEdgeConnected: (state) =>
+        state.pnp.peerConnectionStatus === PEER_CONNECTED
+    })
+  },
+  data () {
+    return { hasSetupSystem: false, isLoading: true }
+  },
+  created () {
+    const setupStatus = window.localStorage.getItem('hasCompletedOnboarding')
```
Comment (ivelin) [inline src/views/Home.vue:124]: @vickywane please use vuex to access local storage items. Shortcutting vuex causes issues with tracking mutations. This is mostly a problem with writes, but we should stick to best practices.

Comment (ivelin) [reply src/views/Home.vue:124]: [See here](https://github.com/ambianic/ambianic-ui/blob/6234856d507f11fa46cf08413e1292515dff52e7/src/views/Settings.vue#L329).

```
    ...mapState({
      edgePeerId: state => state.pnp.remotePeerId,
    }),
```

**Review by ivelin — APPROVED** [review]

File: src/views/Home.vue [hunk 5]
```diff
@@ -1,101 +1,99 @@
...
-            </v-btn>
-          </v-card-actions>
-        </v-card>
-      </v-row>
-      <v-row
-        align="end"
-        justify="center"
-        no-gutters=""
-      >
-        <v-col>
```
[the remark this answers is not in the document]
Comment (vickywane) [reply src/views/Home.vue]: The `UpdateNotifications` component has been restored in recent pushes. This file seems to be outdated

**Pull request merged**